site/app/controllers/admin/ReportController.php

```diff
@@ -652,95 +684,209 @@ public function generateCustomization() {
         }
     }
 
+
+    #[Route("/courses/{_semester}/{_course}/reports/build_form", methods: ['POST'])]
+    public function executeBuildForm(): MultiResponse {
+        // Configure json to go into jobs queue
+        $job_json = (object) [];
+        $job_json->job = 'RunAutoRainbowGrades';
+        $job_json->semester = $this->core->getConfig()->getTerm();
+        $job_json->course = $this->core->getConfig()->getCourse();
+
+        // Encode
+        $job_json = json_encode($job_json, JSON_PRETTY_PRINT);
+
+        // Create path to new jobs queue json
+        $path = '/var/local/submitty/daemon_job_queue/auto_rainbow_' .
+            $this->core->getConfig()->getTerm() .
+            '_' .
+            $this->core->getConfig()->getCourse() .
+            '.json';
+
+        // Place in queue
+        file_put_contents($path, $job_json);
+
+        // Send a success response
+        return new MultiResponse(
+            JsonResponse::getSuccessResponse(['status' => 'success']),
+            null
+        );
+    }
+
+
     #[Route("/courses/{_semester}/{_course}/reports/rainbow_grades_customization/upload", methods: ["POST"])]
     public function uploadRainbowConfig() {
-        $redirect_url =  $this->core->buildCourseUrl((['reports']));
+        $redirect_url = $this->core->buildCourseUrl(['reports']);
         if (empty($_FILES) || !isset($_FILES['config_upload'])) {
             $msg = 'Upload failed: No file to upload';
             $this->core->addErrorMessage($msg);
-            return new MultiResponse(
-                JsonResponse::getErrorResponse($msg),
-                null,
-                new RedirectResponse($redirect_url)
-            );
+            return JsonResponse::getErrorResponse($msg);
         }
 
         $upload = $_FILES['config_upload'];
-        if (empty($upload['tmp_name'])) {
+        if (!isset($upload['tmp_name']) || trim($upload['tmp_name']) === '') {
             $msg = 'Upload failed: Empty tmp name for file';
             $this->core->addErrorMessage($msg);
+            return JsonResponse::getErrorResponse($msg);
+        }
+
+        $rainbow_grades_dir = FileUtils::joinPaths($this->core->getConfig()->getCoursePath(), "rainbow_grades");
+        $destination_path = FileUtils::joinPaths($rainbow_grades_dir, 'manual_customization.json');
+
+        // this is changed from move_uploaded_file to copy because of permission issue (bits not carried over)
+        // copy is expensive, but we are OK because it is small file.
+        // setgid (sticky-bit) gets ignored and doesn't inherit the parent (rainbowgrades dir) permissions
+        // known issue: look https://www.php.net/manual/en/function.move-uploaded-file.php for more details
+        if (!copy($upload['tmp_name'], $destination_path)) {
```

Comment:
Would it be better/easier to just read the file and then write it to the intended location?  That would come with added benefits of ensuring that it's a valid JSON file.

Reply:
I initially liked this idea, but after testing out with existing instructor's customization, `if ($decoded_content === null && json_last_error() !== JSON_ERROR_NONE) {` will fail with syntax error. I think we should leave it like this. Below is what I was trying to do. 
![스크린샷 2024-07-17 오후 1 35 13](https://github.com/user-attachments/assets/b2a0efde-875a-438c-bda7-73a665c696a6)


      $decoded_content = json_decode($json_content);
        if ($decoded_content === null && json_last_error() !== JSON_ERROR_NONE) {
            $error_code = json_last_error();
            $error_message = json_last_error_msg();
            error_log("JSON decode error [$error_code]: $error_message");
            $msg = "Upload failed: Invalid JSON format ($error_message)";
            $this->core->addErrorMessage($msg);
            return JsonResponse::getErrorResponse($msg);
        }